src/rust_inspector/binary.py

```diff
@@ -0,0 +1,278 @@
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/nexB/rust-inspector for support or download.
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+
+import os
+import lief
+import json
+import zlib
+
+from typecode import contenttype
+from typecode.contenttype import get_type
+
+from rust_inspector.blint_binary import parse_symbols
+from rust_inspector.config import SPLIT_CHARACTERS_RUST
+from rust_inspector.config import STANDARD_SYMBOLS_RUST
+
+
+def is_macho(location):
+    """
+    Return True if the file at ``location`` is in macOS/Darwin's Mach-O format, otherwise False.
+    """
+    t = get_type(location)
+    return t.filetype_file.lower().startswith("mach-o") or t.mimetype_file.lower().startswith(
+        "application/x-mach-binary"
+    )
+
+
+def is_executable_binary(location):
+    """
+    Return True if the file at ``location`` is an executable binary.
+    """
+    if not os.path.exists(location):
+        return False
+
+    if not os.path.isfile(location):
+        return False
+
+    typ = contenttype.Type(location)
+
+    if not (typ.is_elf or typ.is_winexe or is_macho(location=location)):
+        return False
+
+    return True
+
+
+def parse_rust_binary(location):
+    """
+    Get a parsed lief._lief.ELF.Binary object from parsing the Rust binary
+    present at `location`.
+    """
+    return lief.parse(location)
+
+
+def get_rust_packages_data(location):
+    """
+    Given a rust binary present at location, parse and return packages
+    data present in the binary.
+
+    This get packages data only if it is present in the rust binary, i.e.
+    if the rust binary was built using `cargo-auditable` which adds a section
+    to the rust binary with data on packages and dependencies.
+    See https://github.com/rust-secure-code/cargo-auditable for more info.
+
+    Code for parsing rust bianries to get package data is from
+    https://github.com/rust-secure-code/cargo-auditable/blob/master/PARSING.md
+    """
+    if not is_executable_binary(location):
+        return
+
+    parsed_binary = parse_rust_binary(location)
+
+    section_deps = None
+    for section in parsed_binary.sections:
+        if section.name == ".dep-v0":
+            section_deps = section
+            break
+
+    decompressed_data = zlib.decompress(section_deps.content)
+    packages_data = json.loads(decompressed_data)
+
+    return packages_data
+
+
+
+def might_have_rust_symbols(string_with_symbols):
+
+    if not string_with_symbols:
+        return False
+
+    if not len(string_with_symbols) > 1:
+        return False
+
+    # Commonly encountered strings which are not rust symbols
+    ignore_strings = ["GCC_except_table", "anonymous", "SAFESEH"]
+    if string_with_symbols in ignore_strings:
+        return False
+
+    # We have sometimes strings which are all uppercase alphabets, these are
```

Comment:
It would be good to have a test or a follow up issue for these cases.

For instance, https://docs.rs/tracing-core/latest/tracing_core/dispatcher/index.html is a thing, and NO_SUBSCRIBER may have been a thing in the past too.

Reply:
We would be doing more improvements on cleaning up symbols, and tuning the symbol matching based on more examples, as a part of https://github.com/aboutcode-org/scancode.io/issues/1475 where we run lot of end-to-end d2d rust examples.